flight/PiOS/posix/pios_serial.c

```diff
@@ -206,48 +206,62 @@ void PIOS_SERIAL_ChangeBaud(uintptr_t serial_id, uint32_t baud)
 		return;
 	}
 
-	/* This serinfo magic for baud rate is deprecated in favor of BOTHER
-	 * on Linux.  However, that requires termios2 and there's various conflict
-	 * that makes it hard to do.
-	 *
-	 * Only consequence right now is an angry kprintf.
-	 */
-
-	if (baud != 9600) {
-		struct serial_struct serinfo;
-
-		if (ioctl(ser_dev->readfd, TIOCGSERIAL, &serinfo)) {
-			perror("ioctl-TIOCGSERIAL");
-		}
-
-		serinfo.flags = (serinfo.flags & ~ASYNC_SPD_MASK) | ASYNC_SPD_CUST;
-		serinfo.custom_divisor = (serinfo.baud_base + (baud / 2)) / baud;
-		uint32_t closest_speed = serinfo.baud_base / serinfo.custom_divisor;
-
-		if ((closest_speed < baud * 99 / 100) ||
-				(closest_speed > baud * 101 / 100)) {
-			printf("Can't attain serial rate %d; using %d\n",
-					baud, closest_speed);
-		}
-
-		if (ioctl(ser_dev->readfd, TIOCSSERIAL, &serinfo)) {
-			perror("ioctl-TIOCSSERIAL");
-		}
-	}
-
 	struct termios options;
 
 	memset(&options, 0, sizeof(options));
 
 	options.c_cflag = CLOCAL | CREAD | CS8;
 
-	if (baud == 9600) {
-		cfsetispeed(&options, B9600);
-		cfsetospeed(&options, B9600);
-	} else {
-		/* The magic above replaces B38400 with the custom rate */
-		cfsetispeed(&options, B38400);
-		cfsetospeed(&options, B38400);
+	switch (baud) {
+		case 1200:
+			printf("Setting Serial ID 0x%x to 1200 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B1200);
+			cfsetospeed(&options, B1200);
+			break;
+		case 2400:
+			printf("Setting Serial ID 0x%x to 2400 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B2400);
+			cfsetospeed(&options, B2400);
+			break;
+		case 4800:
+			printf("Setting Serial ID 0x%x to 4800 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B4800);
+			cfsetospeed(&options, B4800);
+			break;
+		case 9600:
+			printf("Setting Serial ID 0x%x to 9600 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B9600);
+			cfsetospeed(&options, B9600);
+			break;
+		case 19200:
+			printf("Setting Serial ID 0x%x to 19200 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B19200);
+			cfsetospeed(&options, B19200);
+			break;
+		case 38400:
+			printf("Setting Serial ID 0x%x to 38400 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B38400);
+			cfsetospeed(&options, B38400);
+			break;
+		case 57600:
+			printf("Setting Serial ID 0x%x to 57600 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B57600);
+			cfsetospeed(&options, B57600);
+			break;
+		case 115200:
+			printf("Setting Serial ID 0x%x to 115200 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B115200);
+			cfsetospeed(&options, B115200);
+			break;
+		case 230400:
+			printf("Setting Serial ID 0x%x to 230400 Baud\n", (uint32_t)serial_id);
+			cfsetispeed(&options, B230400);
+			cfsetospeed(&options, B230400);
+			break;
+		default:
+			printf("Defaulting Serial ID  0x%x to 9600 Baud\n", (uint32_t)serial_id);
```

Comment:
It's probably good to support all the linux "built in" rates, but this loses the custom_divisor capability / the ability to set e.g. 100kbps for sbus.

The functionality we use on the old code works for me (will reconfirm when I'm back home tomorrow).  Does it complain in dmesg "flightd sets custom speed on ttyS0.  This is deprecated."?  (The functionality is deprecated, but it is in current Linus trees without a universal replacement, and this type of thing is *never* removed because it breaks userspace).

Reply:
Here's what happens with the original pios_serial.c.  Things appear to lock up when trying to set the baud rate for ttyUSB0, I can't even connect to the GCS via IP Network telemetry.

![image](https://user-images.githubusercontent.com/3869782/36646364-e2b4851a-1a33-11e8-8b0f-63d98275008c.png)

But even if I remove the argument for the usb telem radios and try to run, the flightd locks up, that's why I originally went down this path.  

Flightd does not talk to directly to the SBUS receiver from what I see, or is that a future enhancement?

Reply:
> But if I remove the argument for the usb telem radios and try to run, flightd locks up, that's why I originally went down this path.

Without the last -S you can't connect to it by ip address, TCP, over 9000?  It would probably be best to figure this out first.

It makes more sense that this is a USB device that you're not able to set the rate of this way-- almost all of the drivers in linux/drivers/usb/serial support it but some don't.  This is why it'd be good to use the builtin rate constants but still fallback to the current behavior for custom rates (which won't work on USB devices not supporting TIOCSSERIAL)

> Flightd does not talk to directly to the SBUS receiver from what I see, or is that a future enhancement?

There's no code in the sys init for it checked in, but it works now that RTC is somewhat sane in flightd-- I've tried it.